lib/likelihood.c

```diff
@@ -0,0 +1,304 @@
+    } while (msp_increment_edge(ts, edge, parent, child) == 0);
+    if (theta > 0) {
+        *lik = num_mutations * log(total_material * theta) -
+            total_material * theta;
+        for (it = tsk_tree_first(tree); it == 1; it = tsk_tree_next(tree)) {
+            for (j = 0; j < tree->sites_length; j++) {
+                mut = tree->sites[j].mutations[0];
+                ret = tsk_treeseq_get_node(ts, mut.node, child);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_tree_get_parent(tree, child->id, &node_id);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_treeseq_get_node(ts, node_id, parent);
+                if (ret != 0) {
+                    goto out;
+                }
+                branch_length = parent->time - child->time;
+                ret = tsk_tree_get_parent(tree, parent->id, &node_id);
+                while (tsk_tree_get_parent(tree, parent->id, &node_id) == 0
+                        && tree->left_child[parent->id] 
+                        == tree->right_child[parent->id]) {
+                    *child = *parent;
+                    ret = tsk_treeseq_get_node(ts, node_id, parent);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                    branch_length += parent->time - child->time;
+                }
+                ret = tsk_treeseq_get_node(ts, mut.node, child);
+                if (ret != 0) {
+                    goto out;
+                }
+                while (tree->left_child[child->id] >= 0 &&
+                        tree->left_child[child->id] 
+                        == tree->right_child[child->id]) {
+                    *parent = *child;
+                    node_id = tree->left_child[child->id];
+                    ret = tsk_treeseq_get_node(ts, node_id, child);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                    branch_length += parent->time - child->time;
+                }
+                *lik += log(branch_length / total_material);
+            }
+        }
+    } else if (num_mutations > 0) {
+        *lik = -DBL_MAX;
+    } else {
+        *lik = 0;
+    }
+out:
+    if (tree != NULL) {
+        free(tree);
+    }
+    if (edge != NULL) {
+        free(edge);
+    }
+    if (parent != NULL) {
+        free(parent);
+    }
+    if (child != NULL) {
+        free(child);
+    }
+    return ret;
+}
+
+int
+msp_log_likelihood_arg(tsk_treeseq_t *ts, double rho, double *lik) {
+    int ret = 1;
+    int i;
+    int lineages = (int)tsk_treeseq_get_num_samples(ts);
+    tsk_id_t parent_id;
+    double sim_time = 0;
+    double material = lineages * tsk_treeseq_get_sequence_length(ts);
+    double material_in_children;
+    double material_in_parent;
+    double rate;
+    double gap;
+    int first_parent_edge[ts->tables->nodes.num_rows];
```

Comment:
We need to malloc these rather than allocating on the stack. You can only really allocate small bounded things on the stack safely.